compiler/rustc_resolve/src/late/lifetimes.rs

```diff
@@ -1270,106 +1231,127 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
-                    })
-                }
-                &hir::WherePredicate::RegionPredicate(hir::WhereRegionPredicate {
-                    ref lifetime,
-                    bounds,
-                    ..
-                }) => {
-                    self.visit_lifetime(lifetime);
-                    walk_list!(self, visit_param_bound, bounds);
-                }
-                &hir::WherePredicate::EqPredicate(hir::WhereEqPredicate {
-                    ref lhs_ty,
-                    ref rhs_ty,
-                    ..
-                }) => {
-                    self.visit_ty(lhs_ty);
-                    self.visit_ty(rhs_ty);
+                        })
+                    }
+                    &hir::WherePredicate::RegionPredicate(hir::WhereRegionPredicate {
+                        ref lifetime,
+                        bounds,
+                        ..
+                    }) => {
+                        this.visit_lifetime(lifetime);
+                        walk_list!(this, visit_param_bound, bounds);
+                    }
+                    &hir::WherePredicate::EqPredicate(hir::WhereEqPredicate {
+                        ref lhs_ty,
+                        ref rhs_ty,
+                        ..
+                    }) => {
+                        this.visit_ty(lhs_ty);
+                        this.visit_ty(rhs_ty);
+                    }
                 }
             }
-        }
+        })
     }
 
     fn visit_param_bound(&mut self, bound: &'tcx hir::GenericBound<'tcx>) {
+        // FIXME(jackh726): This is pretty weird. `LangItemTrait` doesn't go
+        // through the regular poly trait ref code, so we don't get another
+        // chance to introduce a binder. For now, I'm keeping the existing logic
+        // of "if there isn't a Binder scope above us, add one", but I
+        // imagine there's a better way to go about this.
+        let mut scope = self.scope;
+        let (binders, scope_type) = loop {
+            match scope {
+                Scope::TraitRefBoundary { .. } | Scope::Body { .. } | Scope::Root => {
+                    break (vec![], BinderScopeType::PolyTraitRef);
+                }
+
+                Scope::Binder { hir_id, .. } => {
+                    let binders = self.map.late_bound_vars.entry(*hir_id).or_default().clone();
+                    break (binders, BinderScopeType::Concatenating);
+                }
+
+                Scope::Elision { s, .. }
+                | Scope::ObjectLifetimeDefault { s, .. }
+                | Scope::Supertrait { s, .. } => {
+                    scope = s;
+                }
+            }
+        };
         match bound {
-            hir::GenericBound::LangItemTrait(_, _, hir_id, _) if self.trait_ref_hack.is_none() => {
-                self.map.late_bound_vars.insert(*hir_id, vec![]);
+            hir::GenericBound::LangItemTrait(_, _, hir_id, _) => {
+                self.map.late_bound_vars.insert(*hir_id, binders);
```

Comment:
Except that I don't quite get why we need any code here at *all*. It seems like the `LangItemTrait` is basically a "poly trait ref" that binds no variables at all. We could probably pull out a helper from `visit_poly_trait_ref` that does the exact same setup code -- which would be good -- or else delete this code altogether, because it's not really needed. I can't quite see any case where introducing a `Binder` here changes behavior. 

I guess the only change could be that there is a `PolyTraitRef` boundary (with no variables) instead of a `TraitRefBoundary` binder. Are those equivalent? I would think so (does that suggest a further simplification is possible?).